compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3144,6 +3144,125 @@ static inline void loadArrayCmpSources(TR::Node *node, TR::InstOpCode::Mnemonic
+
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, indexReg, 0);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi8, node, condReg, tempReg, 16);
+
+   // We don't need length anymore as we can calculate the appropriate index by using indexReg and the remainder
+   generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, returnReg, tempReg, 0, 0xF);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::blt, node, residueStartLabel, condReg);
+
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::srawi, node, tempReg, tempReg, 4);
+   generateSrc1Instruction(cg, TR::InstOpCode::mtctr, node, tempReg);
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, loopStartLabel);
+   startLabel->setStartInternalControlFlow();
+
+   // main-loop
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvb16x, node, vec0Reg, indexReg, src1AddrReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvb16x, node, vec1Reg, indexReg, src2AddrReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vcmpequb_r, node, vec0Reg, vec0Reg, vec1Reg);
+
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bge, node, resultLabel, condReg);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, indexReg, indexReg, 16);
+
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bdnz, node, loopStartLabel, condReg);
+   // main-loop end
+
+   // residue start
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, residueStartLabel);
+
+   generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldicr, node, temp2Reg, returnReg, 56, 0xFF00000000000000);   
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, tempReg, src1AddrReg, indexReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvll, node, vec0Reg, tempReg, temp2Reg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, tempReg, src2AddrReg, indexReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvll, node, vec1Reg, tempReg, temp2Reg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vcmpequb, node, vec0Reg, vec0Reg, vec1Reg);
+   // residue end
+
+   // result
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, resultLabel);
+
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vnor, node, vec0Reg, vec0Reg, vec0Reg);
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::vclzlsbb, node, temp2Reg, vec0Reg);
+   
+   // vec-eq-mask = -1 (if equal) | 0 (it not-equal)
+   generateShiftLeftImmediateLong(cg, node, tempReg, temp2Reg, 59);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::sradi, node, tempReg, tempReg, 63);
+
+   // offset = matched-byte-count & ~(vec-eq-mask) + remainder & (vec-eq-mask)
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::andc, node, temp2Reg, temp2Reg, tempReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::AND, node, returnReg, returnReg, tempReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, returnReg, returnReg, temp2Reg);
```

Comment:
One of my wip implementation was with `isel`. As we need to subtract 1 from length in case of arrays being equal, later changed the implementation to use the mask instead of `isel`.

Reply:
Ah, I didn't notice that. However, it looks like this is only necessary when `node->isArrayCmpLen()` is false. Maybe a better solution would be to split the implementation here and use `iseleq` when getting that -1 on a match is not required, only using the masks if required?

Reply:
I changed to my old implementation, kindly review again.